File: src/insulindian_miracle/policies.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
import math
from typing import Protocol

import numpy as np

from .model import Action, SimulationConfig, SiteStateSnapshot, base_geography, compute_reward, network_bonus, sigmoid


class Policy(Protocol):
    name: str

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        ...

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        ...


CONTEXTUAL_FEATURE_DIM = 11


def _contextual_feature_vector(index: int, states: list[SiteStateSnapshot], config: SimulationConfig) -> np.ndarray:
    state = states[index]
    # For Phase 1, we use a simplified vector since we don't have the full Site object in snapshot
    return np.asarray(
        [
            1.0,
            0.0, # base_geography - could be added to snapshot if needed
            state.resource_rent,
            state.extraction,
            state.openness,
            state.adaptability,
            state.productive_capital,
            math.log1p(max(state.population, 1)),
            0.0, # network_bonus
            float(state.boomtown),
            float(state.trade_cluster),
        ],
        dtype=float,
    )


def _contextual_feature_matrix(states: list[SiteStateSnapshot], config: SimulationConfig) -> np.ndarray:
    if not states:
        return np.zeros((0, CONTEXTUAL_FEATURE_DIM), dtype=float)
    return np.vstack([_contextual_feature_vector(index, states, config) for index in range(len(states))])


@dataclass(slots=True)
class EpsilonGreedyPolicy:
    arm_count: int
    seed: int
    epsilon: float = 0.1
    name: str = "epsilon-greedy"
    rng: np.random.Generator = field(init=False, repr=False)
    counts: np.ndarray = field(init=False, repr=False)
    values: np.ndarray = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.rng = np.random.default_rng(self.seed)
        self.counts = np.zeros(self.arm_count, dtype=int)
        self.values = np.zeros(self.arm_count, dtype=float)

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        unseen = np.where(self.counts == 0)[0]
        if unseen.size:
            return int(unseen[0])
        if self.rng.random() < self.epsilon:
            return int(self.rng.integers(0, self.arm_count))
        return int(np.argmax(self.values))

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        self.counts[chosen_arm] += 1
        count = self.counts[chosen_arm]
        self.values[chosen_arm] += (reward - self.values[chosen_arm]) / count


@dataclass(slots=True)
class UCB1Policy:
    arm_count: int
    seed: int
    exploration: float = 2.0
    name: str = "ucb1"
    counts: np.ndarray = field(init=False, repr=False)
    values: np.ndarray = field(init=False, repr=False)
    steps: int = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.counts = np.zeros(self.arm_count, dtype=int)
        self.values = np.zeros(self.arm_count, dtype=float)
        self.steps = 0

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        unseen = np.where(self.counts == 0)[0]
        if unseen.size:
            return int(unseen[0])
        bonus = np.sqrt((self.exploration * np.log(max(self.steps, 1))) / self.counts)
        return int(np.argmax(self.values + bonus))

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        self.steps += 1
        self.counts[chosen_arm] += 1
        count = self.counts[chosen_arm]
        self.values[chosen_arm] += (reward - self.values[chosen_arm]) / count


@dataclass(slots=True)
class GaussianThompsonPolicy:
    arm_count: int
    seed: int
    observation_variance: float = 9.0
    prior_mean: float = 0.0
    prior_variance: float = 16.0
    posterior_decay: float = 1.0
    minimum_exploration_variance: float = 0.3
    name: str = "gaussian-thompson"
    rng: np.random.Generator = field(init=False, repr=False)
    precisions: np.ndarray = field(init=False, repr=False)
    mean_precision: np.ndarray = field(init=False, repr=False)
    counts: np.ndarray = field(init=False, repr=False)
    prior_precision: float = field(init=False, repr=False)
    prior_mean_precision: float = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.rng = np.random.default_rng(self.seed)
        self.prior_precision = 1.0 / self.prior_variance
        self.prior_mean_precision = self.prior_mean / self.prior_variance
        self.precisions = np.full(self.arm_count, self.prior_precision)
        self.mean_precision = np.full(self.arm_count, self.prior_mean_precision)
        self.counts = np.zeros(self.arm_count, dtype=float)

    def _sample_arm(self) -> int:
        variances = np.maximum(1.0 / self.precisions, self.minimum_exploration_variance / np.sqrt(self.counts + 1.0))
        means = self.mean_precision / self.precisions
        draws = self.rng.normal(means, np.sqrt(variances))
        return int(np.argmax(draws))

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        unseen = np.where(self.counts < 0.5)[0]
        if unseen.size:
            return int(unseen[0])
        return self._sample_arm()

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        obs_precision = 1.0 / self.observation_variance
        self.counts *= self.posterior_decay
        self.counts[chosen_arm] += 1.0
        self.precisions = self.prior_precision + self.posterior_decay * (self.precisions - self.prior_precision)
        self.mean_precision = self.prior_mean_precision + self.posterior_decay * (self.mean_precision - self.prior_mean_precision)
        self.precisions[chosen_arm] += obs_precision
        self.mean_precision[chosen_arm] += reward * obs_precision


@dataclass(slots=True)
class DiscountedUCBPolicy:
    arm_count: int
    seed: int
    gamma: float = 0.97
    exploration: float = 2.0
    name: str = "discounted-ucb"
    counts: np.ndarray = field(init=False, repr=False)
    reward_sums: np.ndarray = field(init=False, repr=False)
    total_mass: float = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.gamma = min(max(self.gamma, 0.0), 1.0)
        self.counts = np.zeros(self.arm_count, dtype=float)
        self.reward_sums = np.zeros(self.arm_count, dtype=float)
        self.total_mass = 0.0

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        unseen = np.where(self.counts <= 1e-9)[0]
        if unseen.size:
            return int(unseen[0])
        means = self.reward_sums / np.maximum(self.counts, 1e-9)
        bonus = np.sqrt((self.exploration * np.log(self.total_mass + 1.0)) / np.maximum(self.counts, 1e-9))
        return int(np.argmax(means + bonus))

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        self.counts *= self.gamma
        self.reward_sums *= self.gamma
        self.total_mass = self.total_mass * self.gamma + 1.0
        self.counts[chosen_arm] += 1.0
        self.reward_sums[chosen_arm] += reward


@dataclass(slots=True)
class SlidingWindowUCBPolicy:
    arm_count: int
    seed: int
    window_size: int = 40
    exploration: float = 2.0
    name: str = "sliding-window-ucb"
    observation_window: deque[tuple[int, float]] = field(init=False, repr=False)
    counts: np.ndarray = field(init=False, repr=False)
    reward_sums: np.ndarray = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.window_size = max(1, int(self.window_size))
        self.observation_window = deque()
        self.counts = np.zeros(self.arm_count, dtype=float)
        self.reward_sums = np.zeros(self.arm_count, dtype=float)

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        unseen = np.where(self.counts <= 1e-9)[0]
        if unseen.size:
            return int(unseen[0])
        means = self.reward_sums / np.maximum(self.counts, 1.0)
        total_mass = max(float(len(self.observation_window)), 1.0)
        bonus = np.sqrt((self.exploration * np.log(total_mass + 1.0)) / np.maximum(self.counts, 1.0))
        return int(np.argmax(means + bonus))

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        if len(self.observation_window) == self.window_size:
            expired_arm, expired_reward = self.observation_window.popleft()
            self.counts[expired_arm] = max(self.counts[expired_arm] - 1.0, 0.0)
            self.reward_sums[expired_arm] -= expired_reward
            if self.counts[expired_arm] <= 1e-9:
                self.counts[expired_arm] = 0.0
                self.reward_sums[expired_arm] = 0.0

        reward_value = float(reward)
        self.observation_window.append((chosen_arm, reward_value))
        self.counts[chosen_arm] += 1.0
        self.reward_sums[chosen_arm] += reward_value


@dataclass(slots=True)
class DiscountedGaussianThompsonPolicy(GaussianThompsonPolicy):
    name: str = "discounted-gaussian-thompson"
    observed: np.ndarray = field(init=False, repr=False)

    def __post_init__(self) -> None:
        super().__post_init__()
        self.observed = np.zeros(self.arm_count, dtype=bool)

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        unseen = np.where(~self.observed)[0]
        if unseen.size:
            return int(unseen[0])
        return self._sample_arm()

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        super().update(chosen_arm, reward, states)
        self.observed[chosen_arm] = True


@dataclass(slots=True)
class LinUCBPolicy:
    arm_count: int
    seed: int
    config: SimulationConfig
    alpha: float = 1.15
    ridge: float = 1.0
    name: str = "linucb"
    covariance: np.ndarray = field(init=False, repr=False)
    reward_vector: np.ndarray = field(init=False, repr=False)
    last_features: np.ndarray = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.ridge = max(self.ridge, 1e-6)
        self.covariance = self.ridge * np.eye(CONTEXTUAL_FEATURE_DIM, dtype=float)
        self.reward_vector = np.zeros(CONTEXTUAL_FEATURE_DIM, dtype=float)
        self.last_features = np.zeros((self.arm_count, CONTEXTUAL_FEATURE_DIM), dtype=float)

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        features = _contextual_feature_matrix(states, self.config)
        self.last_features = features
        inverse = np.linalg.inv(self.covariance)
        theta = inverse @ self.reward_vector
        bonuses = np.sqrt(np.einsum("ij,jk,ik->i", features, inverse, features))
        scores = features @ theta + self.alpha * bonuses
        return int(np.argmax(scores))

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        feature = self.last_features[chosen_arm]
        self.covariance += np.outer(feature, feature)
        self.reward_vector += reward * feature


@dataclass(slots=True)
class LinearThompsonPolicy:
    arm_count: int
    seed: int
    config: SimulationConfig
    ridge: float = 1.0
    observation_variance: float = 9.0
    sampling_scale: float = 1.0
    name: str = "linear-thompson"
    rng: np.random.Generator = field(init=False, repr=False)
    precision: np.ndarray = field(init=False, repr=False)
    reward_precision: np.ndarray = field(init=False, repr=False)
    last_features: np.ndarray = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.rng = np.random.default_rng(self.seed)
        self.ridge = max(self.ridge, 1e-6)
        self.observation_variance = max(self.observation_variance, 1e-6)
        self.sampling_scale = max(self.sampling_scale, 1e-6)
        self.precision = self.ridge * np.eye(CONTEXTUAL_FEATURE_DIM, dtype=float)
        self.reward_precision = np.zeros(CONTEXTUAL_FEATURE_DIM, dtype=float)
        self.last_features = np.zeros((self.arm_count, CONTEXTUAL_FEATURE_DIM), dtype=float)

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        features = _contextual_feature_matrix(states, self.config)
        self.last_features = features
        covariance = np.linalg.inv(self.precision)
        mean = covariance @ self.reward_precision
        chol = np.linalg.cholesky(covariance + 1e-9 * np.eye(CONTEXTUAL_FEATURE_DIM, dtype=float))
        theta = mean + self.sampling_scale * (chol @ self.rng.normal(size=CONTEXTUAL_FEATURE_DIM))
        scores = features @ theta
        return int(np.argmax(scores))

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        feature = self.last_features[chosen_arm]
        obs_precision = 1.0 / self.observation_variance
        self.precision += obs_precision * np.outer(feature, feature)
        self.reward_precision += obs_precision * reward * feature


@dataclass(slots=True)
class WhittleIndexPolicy:
    arm_count: int
    seed: int
    config: SimulationConfig
    name: str = "whittle-index"
    # Simplified Whittle for Phase 1 - focused on separation not full logic porting
    def __post_init__(self) -> None:
        pass

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        # Dummy implementation for now to keep the code running
        return 0

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        pass


@dataclass(slots=True)
class MyopicOraclePolicy:
    arm_count: int
    seed: int
    config: SimulationConfig
    name: str = "myopic-oracle"

    def __post_init__(self) -> None:
        pass

    def select_site(self, states: list[SiteStateSnapshot]) -> Action:
        # Simplified: choose site with highest resource rent as a proxy for myopic gain
        rents = [s.resource_rent for s in states]
        return int(np.argmax(rents))

    def update(self, chosen_arm: Action, reward: float, states: list[SiteStateSnapshot]) -> None:
        pass


def build_policy(name: str, arm_count: int, seed: int, config: SimulationConfig) -> Policy:
    normalized = name.lower()
    if normalized in {"epsilon", "epsilon-greedy", "eps-greedy"}:
        return EpsilonGreedyPolicy(arm_count=arm_count, seed=seed)
    if normalized in {"ucb", "ucb1"}:
        return UCB1Policy(arm_count=arm_count, seed=seed)
    if normalized in {"discounted-ucb", "ducb"}:
        return DiscountedUCBPolicy(arm_count=arm_count, seed=seed, gamma=config.discounted_ucb_gamma)
    if normalized in {"sliding-window-ucb", "sliding-ucb", "sw-ucb"}:
        return SlidingWindowUCBPolicy(arm_count=arm_count, seed=seed, window_size=config.sliding_window_ucb_window)
    if normalized in {"ts", "thompson", "gaussian-thompson"}:
        return GaussianThompsonPolicy(arm_count=arm_count, seed=seed, posterior_decay=config.thompson_posterior_decay)
    if normalized in {"discounted-gaussian-thompson", "discounted-thompson", "discounted-ts"}:
        return DiscountedGaussianThompsonPolicy(
            arm_count=arm_count,
            seed=seed,
            posterior_decay=config.discounted_thompson_posterior_decay,
        )
    if normalized in {"linucb", "linear-ucb"}:
        return LinUCBPolicy(
            arm_count=arm_count,
            seed=seed,
            config=config,
            alpha=config.linucb_alpha,
            ridge=config.linear_bandit_ridge,
        )
    if normalized in {"linear-thompson", "linear-ts"}:
        return LinearThompsonPolicy(
            arm_count=arm_count,
            seed=seed,
            config=config,
            ridge=config.linear_bandit_ridge,
            observation_variance=config.linear_thompson_observation_variance,
            sampling_scale=config.linear_thompson_sampling_scale,
        )
    if normalized in {"whittle", "whittle-index"}:
        return WhittleIndexPolicy(arm_count=arm_count, seed=seed, config=config)
    if normalized in {"oracle", "myopic-oracle"}:
        return MyopicOraclePolicy(arm_count=arm_count, seed=seed, config=config)
    raise ValueError(f"Unknown policy: {name}")

File: rust/src/policies.rs
use crate::snapshot_physics::network_bonus_snapshot;
use crate::types::{SimulationConfig, SiteStateSnapshot};
use ndarray::Array1;
use rand::Rng;
use rand_distr::{Normal, Distribution};
use std::collections::VecDeque;

pub trait Policy {
    fn select_site(&mut self, states: &[SiteStateSnapshot]) -> usize;
    fn update(&mut self, chosen_arm: usize, reward: f64, states: &[SiteStateSnapshot]);
    fn wants_snapshots(&self) -> bool { false }
}

pub struct EpsilonGreedyPolicy<R: Rng> {
    arm_count: usize,
    rng: R,
    epsilon: f64,
    counts: Array1<usize>,
    values: Array1<f64>,
}

impl<R: Rng> EpsilonGreedyPolicy<R> {
    pub fn new(arm_count: usize, rng: R, epsilon: f64) -> Self {
        Self {
            arm_count,
            rng,
            epsilon,
            counts: Array1::zeros(arm_count),
            values: Array1::zeros(arm_count),
        }
    }
}

impl<R: Rng> Policy for EpsilonGreedyPolicy<R> {
    fn select_site(&mut self, _states: &[SiteStateSnapshot]) -> usize {
        for i in 0..self.arm_count {
            if self.counts[i] == 0 {
                return i;
            }
        }
        if self.rng.gen::<f64>() < self.epsilon {
            self.rng.gen_range(0..self.arm_count)
        } else {
            let mut best_arm = 0;
            let mut best_val = f64::NEG_INFINITY;
            for i in 0..self.arm_count {
                if self.values[i] > best_val {
                    best_val = self.values[i];
                    best_arm = i;
                }
            }
            best_arm
        }
    }

    fn update(&mut self, chosen_arm: usize, reward: f64, _states: &[SiteStateSnapshot]) {
        self.counts[chosen_arm] += 1;
        let count = self.counts[chosen_arm] as f64;
        self.values[chosen_arm] += (reward - self.values[chosen_arm]) / count;
    }
}

pub struct UCB1Policy {
    arm_count: usize,
    exploration: f64,
    counts: Array1<usize>,
    values: Array1<f64>,
    steps: usize,
}

impl UCB1Policy {
    pub fn new(arm_count: usize, exploration: f64) -> Self {
        Self {
            arm_count,
            exploration,
            counts: Array1::zeros(arm_count),
            values: Array1::zeros(arm_count),
            steps: 0,
        }
    }
}

impl Policy for UCB1Policy {
    fn select_site(&mut self, _states: &[SiteStateSnapshot]) -> usize {
        for i in 0..self.arm_count {
            if self.counts[i] == 0 {
                return i;
            }
        }
        
        let mut best_arm = 0;
        let mut best_val = f64::NEG_INFINITY;
        let log_steps = (self.steps.max(1) as f64).ln();
        
        for i in 0..self.arm_count {
            let bonus = (self.exploration * log_steps / self.counts[i] as f64).sqrt();
            let score = self.values[i] + bonus;
            if score > best_val {
                best_val = score;
                best_arm = i;
            }
        }
        best_arm
    }

    fn update(&mut self, chosen_arm: usize, reward: f64, _states: &[SiteStateSnapshot]) {
        self.steps += 1;
        self.counts[chosen_arm] += 1;
        let count = self.counts[chosen_arm] as f64;
        self.values[chosen_arm] += (reward - self.values[chosen_arm]) / count;
    }
}

pub struct GaussianThompsonPolicy<R: Rng> {
    arm_count: usize,
    rng: R,
    observation_variance: f64,
    prior_precision: f64,
    prior_mean_precision: f64,
    posterior_decay: f64,
    minimum_exploration_variance: f64,
    precisions: Array1<f64>,
    mean_precision: Array1<f64>,
    counts: Array1<f64>,
    observed: Array1<bool>,
    discounted: bool,
}

impl<R: Rng> GaussianThompsonPolicy<R> {
    pub fn new(
        arm_count: usize,
        rng: R,
        observation_variance: f64,
        prior_mean: f64,
        prior_variance: f64,
        posterior_decay: f64,
        minimum_exploration_variance: f64,
        discounted: bool,
    ) -> Self {
        let prior_precision = 1.0 / prior_variance;
        let prior_mean_precision = prior_mean / prior_variance;
        
        Self {
            arm_count,
            rng,
            observation_variance,
            prior_precision,
            prior_mean_precision,
            posterior_decay,
            minimum_exploration_variance,
            precisions: Array1::from_elem(arm_count, prior_precision),
            mean_precision: Array1::from_elem(arm_count, prior_mean_precision),
            counts: Array1::zeros(arm_count),
            observed: Array1::from_elem(arm_count, false),
            discounted,
        }
    }
    
    fn sample_arm(&mut self) -> usize {
        let mut best_arm = 0;
        let mut best_val = f64::NEG_INFINITY;
        
        for i in 0..self.arm_count {
            let var1 = 1.0 / self.precisions[i];
            let var2 = self.minimum_exploration_variance / (self.counts[i] + 1.0).sqrt();
            let variance = var1.max(var2);
            let mean = self.mean_precision[i] / self.precisions[i];
            
            let normal = Normal::new(mean, variance.sqrt()).unwrap();
            let draw = normal.sample(&mut self.rng);
            if draw > best_val {
                best_val = draw;
                best_arm = i;
            }
        }
        best_arm
    }
}

impl<R: Rng> Policy for GaussianThompsonPolicy<R> {
    fn select_site(&mut self, _states: &[SiteStateSnapshot]) -> usize {
        if self.discounted {
            for i in 0..self.arm_count {
                if !self.observed[i] {
                    return i;
                }
            }
        } else {
            for i in 0..self.arm_count {
                if self.counts[i] < 0.5 {
                    return i;
                }
            }
        }
        self.sample_arm()
    }

    fn update(&mut self, chosen_arm: usize, reward: f64, _states: &[SiteStateSnapshot]) {
        let obs_precision = 1.0 / self.observation_variance;
        
        for i in 0..self.arm_count {
            self.counts[i] *= self.posterior_decay;
            self.precisions[i] = self.prior_precision + self.posterior_decay * (self.precisions[i] - self.prior_precision);
            self.mean_precision[i] = self.prior_mean_precision + self.posterior_decay * (self.mean_precision[i] - self.prior_mean_precision);
        }
        
        self.counts[chosen_arm] += 1.0;
        self.precisions[chosen_arm] += obs_precision;
        self.mean_precision[chosen_arm] += reward * obs_precision;
        
        if self.discounted {
            self.observed[chosen_arm] = true;
        }
    }
}

pub struct DiscountedUCBPolicy {
    arm_count: usize,
    gamma: f64,
    exploration: f64,
    counts: Array1<f64>,
    reward_sums: Array1<f64>,
    total_mass: f64,
}

impl DiscountedUCBPolicy {
    pub fn new(arm_count: usize, gamma: f64, exploration: f64) -> Self {
        Self {
            arm_count,
            gamma: gamma.clamp(0.0, 1.0),
            exploration,
            counts: Array1::zeros(arm_count),
            reward_sums: Array1::zeros(arm_count),
            total_mass: 0.0,
        }
    }
}

impl Policy for DiscountedUCBPolicy {
    fn select_site(&mut self, _states: &[SiteStateSnapshot]) -> usize {
        for i in 0..self.arm_count {
            if self.counts[i] <= 1e-9 {
                return i;
            }
        }
        
        let mut best_arm = 0;
        let mut best_val = f64::NEG_INFINITY;
        let log_mass = (self.total_mass + 1.0).ln();
        
        for i in 0..self.arm_count {
            let mean = self.reward_sums[i] / self.counts[i].max(1e-9);
            let bonus = (self.exploration * log_mass / self.counts[i].max(1e-9)).sqrt();
            let score = mean + bonus;
            if score > best_val {
                best_val = score;
                best_arm = i;
            }
        }
        best_arm
    }

    fn update(&mut self, chosen_arm: usize, reward: f64, _states: &[SiteStateSnapshot]) {
        for i in 0..self.arm_count {
            self.counts[i] *= self.gamma;
            self.reward_sums[i] *= self.gamma;
        }
        self.total_mass = self.total_mass * self.gamma + 1.0;
        
        self.counts[chosen_arm] += 1.0;
        self.reward_sums[chosen_arm] += reward;
    }
}

pub struct SlidingWindowUCBPolicy {
    arm_count: usize,
    window_size: usize,
    exploration: f64,
    observation_window: VecDeque<(usize, f64)>,
    counts: Array1<f64>,
    reward_sums: Array1<f64>,
}

impl SlidingWindowUCBPolicy {
    pub fn new(arm_count: usize, window_size: usize, exploration: f64) -> Self {
        Self {
            arm_count,
            window_size: window_size.max(1),
            exploration,
            observation_window: VecDeque::new(),
            counts: Array1::zeros(arm_count),
            reward_sums: Array1::zeros(arm_count),
        }
    }
}

impl Policy for SlidingWindowUCBPolicy {
    fn select_site(&mut self, _states: &[SiteStateSnapshot]) -> usize {
        for i in 0..self.arm_count {
            if self.counts[i] <= 1e-9 {
                return i;
            }
        }
        
        let mut best_arm = 0;
        let mut best_val = f64::NEG_INFINITY;
        let total_mass = (self.observation_window.len() as f64).max(1.0);
        let log_mass = (total_mass + 1.0).ln();
        
        for i in 0..self.arm_count {
            let mean = self.reward_sums[i] / self.counts[i].max(1.0);
            let bonus = (self.exploration * log_mass / self.counts[i].max(1.0)).sqrt();
            let score = mean + bonus;
            if score > best_val {
                best_val = score;
                best_arm = i;
            }
        }
        best_arm
    }

    fn update(&mut self, chosen_arm: usize, reward: f64, _states: &[SiteStateSnapshot]) {
        if self.observation_window.len() == self.window_size {
            if let Some((expired_arm, expired_reward)) = self.observation_window.pop_front() {
                self.counts[expired_arm] = (self.counts[expired_arm] - 1.0).max(0.0);
                self.reward_sums[expired_arm] -= expired_reward;
                if self.counts[expired_arm] <= 1e-9 {
                    self.counts[expired_arm] = 0.0;
                    self.reward_sums[expired_arm] = 0.0;
                }
            }
        }
        
        self.observation_window.push_back((chosen_arm, reward));
        self.counts[chosen_arm] += 1.0;
        self.reward_sums[chosen_arm] += reward;
    }
}

pub struct MyopicOraclePolicy {
}

impl MyopicOraclePolicy {
    pub fn new(_arm_count: usize) -> Self {
        Self { }
    }
}

impl Policy for MyopicOraclePolicy {
    fn wants_snapshots(&self) -> bool { true }
    fn select_site(&mut self, states: &[SiteStateSnapshot]) -> usize {
        let mut best_arm = 0;
        let mut best_rent = f64::NEG_INFINITY;
        for (i, state) in states.iter().enumerate() {
            if state.resource_rent > best_rent {
                best_rent = state.resource_rent;
                best_arm = i;
            }
        }
        best_arm
    }
    fn update(&mut self, _chosen_arm: usize, _reward: f64, _states: &[SiteStateSnapshot]) {}
}

const CONTEXTUAL_FEATURE_DIM: usize = 11;

fn get_contextual_feature_vector(
    index: usize,
    states: &[SiteStateSnapshot],
    config: &SimulationConfig,
) -> Array1<f64> {
    let state = &states[index];
    let nb = network_bonus_snapshot(index, states, config).clamp(0.0, 3.0);
    ndarray::array![
        1.0,
        state.geography,
        state.resource_rent,
        state.extraction,
        state.openness,
        state.adaptability,
        state.productive_capital,
        (state.population.max(1) as f64).ln_1p(),
        nb,
        if state.boomtown { 1.0 } else { 0.0 },
        if state.trade_cluster { 1.0 } else { 0.0 }
    ]
}

use crate::math_utils::{invert_matrix, cholesky_decomposition};
use ndarray::Array2;

pub struct LinUCBPolicy {
    arm_count: usize,
    alpha: f64,
    config: SimulationConfig,
    inv_covariance: Array2<f64>,
    reward_vector: Array1<f64>,
    last_features: Vec<Array1<f64>>,
}

impl LinUCBPolicy {
    pub fn new(arm_count: usize, alpha: f64, ridge: f64, config: SimulationConfig) -> Self {
        let ridge_val = ridge.max(1e-6);
        let inv_covariance = Array2::eye(CONTEXTUAL_FEATURE_DIM) * (1.0 / ridge_val);
        Self {
            arm_count,
            alpha,
            config,
            inv_covariance,
            reward_vector: Array1::zeros(CONTEXTUAL_FEATURE_DIM),
            last_features: vec![Array1::zeros(CONTEXTUAL_FEATURE_DIM); arm_count],
        }
    }
}

impl Policy for LinUCBPolicy {
    fn wants_snapshots(&self) -> bool { true }
    fn select_site(&mut self, states: &[SiteStateSnapshot]) -> usize {
        let theta = self.inv_covariance.dot(&self.reward_vector);
        
        let mut best_arm = 0;
        let mut best_val = f64::NEG_INFINITY;
        
        for i in 0..self.arm_count {
            let feat = get_contextual_feature_vector(i, states, &self.config);
            self.last_features[i] = feat.clone();
            
            let temp = self.inv_covariance.dot(&feat);
            let variance = feat.dot(&temp);
            let bonus = self.alpha * variance.sqrt();
            let score = feat.dot(&theta) + bonus;
            
            if score > best_val {
                best_val = score;
                best_arm = i;
            }
        }
        best_arm
    }

    fn update(&mut self, chosen_arm: usize, reward: f64, _states: &[SiteStateSnapshot]) {
        let x = &self.last_features[chosen_arm];
        
        // Sherman-Morrison rank-1 update for inverse:
        // (A + xx^T)^{-1} = A^{-1} - (A^{-1}xx^T A^{-1}) / (1 + x^T A^{-1}x)
        let a_inv_x = self.inv_covariance.dot(x);
        let x_t_a_inv_x = x.dot(&a_inv_x);
        let denom = 1.0 + x_t_a_inv_x;
        
        for i in 0..CONTEXTUAL_FEATURE_DIM {
            for j in 0..CONTEXTUAL_FEATURE_DIM {
                self.inv_covariance[[i, j]] -= (a_inv_x[i] * a_inv_x[j]) / denom;
            }
            self.reward_vector[i] += reward * x[i];
        }
    }
}

pub struct LinearThompsonPolicy<R: Rng> {
    arm_count: usize,
    rng: R,
    config: SimulationConfig,
    observation_variance: f64,
    sampling_scale: f64,
    precision: Array2<f64>,
    reward_precision: Array1<f64>,
    last_features: Vec<Array1<f64>>,
}

impl<R: Rng> LinearThompsonPolicy<R> {
    pub fn new(
        arm_count: usize,
        rng: R,
        ridge: f64,
        observation_variance: f64,
        sampling_scale: f64,
        config: SimulationConfig,
    ) -> Self {
        let ridge_val = ridge.max(1e-6);
        let precision = Array2::eye(CONTEXTUAL_FEATURE_DIM) * ridge_val;
        Self {
            arm_count,
            rng,
            config,
            observation_variance: observation_variance.max(1e-6),
            sampling_scale: sampling_scale.max(1e-6),
            precision,
            reward_precision: Array1::zeros(CONTEXTUAL_FEATURE_DIM),
            last_features: vec![Array1::zeros(CONTEXTUAL_FEATURE_DIM); arm_count],
        }
    }
}

impl<R: Rng> Policy for LinearThompsonPolicy<R> {
    fn wants_snapshots(&self) -> bool { true }
    fn select_site(&mut self, states: &[SiteStateSnapshot]) -> usize {
        let mut covariance = invert_matrix(&self.precision).unwrap_or_else(|| Array2::eye(CONTEXTUAL_FEATURE_DIM));
        let mean = covariance.dot(&self.reward_precision);
        
        for i in 0..CONTEXTUAL_FEATURE_DIM {
            covariance[[i, i]] += 1e-9;
        }
        
        let chol = cholesky_decomposition(&covariance).unwrap_or_else(|| Array2::eye(CONTEXTUAL_FEATURE_DIM));
        
        let mut normal_sample = Array1::<f64>::zeros(CONTEXTUAL_FEATURE_DIM);
        let normal = Normal::new(0.0, 1.0).unwrap();
        for i in 0..CONTEXTUAL_FEATURE_DIM {
            normal_sample[i] = normal.sample(&mut self.rng);
        }
        
        let theta = &mean + &(chol.dot(&normal_sample) * self.sampling_scale);
        
        let mut best_arm = 0;
        let mut best_val = f64::NEG_INFINITY;
        
        for i in 0..self.arm_count {
            let feat = get_contextual_feature_vector(i, states, &self.config);
            self.last_features[i] = feat.clone();
            
            let score = feat.dot(&theta);
            if score > best_val {
                best_val = score;
                best_arm = i;
            }
        }
        best_arm
    }

    fn update(&mut self, chosen_arm: usize, reward: f64, _states: &[SiteStateSnapshot]) {
        let feat = &self.last_features[chosen_arm];
        let obs_precision = 1.0 / self.observation_variance;
        for i in 0..CONTEXTUAL_FEATURE_DIM {
            for j in 0..CONTEXTUAL_FEATURE_DIM {
                self.precision[[i, j]] += obs_precision * feat[i] * feat[j];
            }
            self.reward_precision[i] += obs_precision * reward * feat[i];
        }
    }
}

